Implement waitset handle

jacquelinekay opened the pull request:
Connects to ros2/rmw#49

jacquelinekay commented:
here's something pretty cool:

when I run the allocator test with --all-tokens on the waitset_handle set of branches (meaning no functions are ignored from the middleware), the connext version of the test passes!

dirk-thomas commented on rmw_connext_shared_cpp/src/shared_functions.cpp:
```diff
@@ -510,6 +510,185 @@ destroy_guard_condition(const char * implementation_identifier,
+  // From here onward, error results in unrolling in the goto fail block.
+  if (!waitset) {
+    RMW_SET_ERROR_MSG("failed to allocate waitset");
+    goto fail;
```
In this case `waitset_info` is a `nullptr` but in the first line under `fail` `waitset_info->active_conditions` is being accessed.

jacquelinekay replied:
757002b8282f733a93f074c8352f059ed7327c8b will fix this by checking the pointer in the fail block.

jacquelinekay commented:
I believe this and ros2/rmw_opensplice#98 are the last pull requests that need approval and that I have addressed all comments. @dirk-thomas what do you think?

dirk-thomas commented:
lgtm.
